Mathlib/LinearAlgebra/StdBasis.lean

```diff
@@ -0,0 +1,310 @@
+  single
+#align linear_map.std_basis LinearMap.stdBasis
+
+theorem stdBasis_apply (i : ι) (b : φ i) : stdBasis R φ i b = update (0 : (a : ι) → φ a) i b :=
+  rfl
+#align linear_map.std_basis_apply LinearMap.stdBasis_apply
+
+@[simp]
+theorem stdBasis_apply' (i i' : ι) : (stdBasis R (fun _x : ι => R) i) 1 i' = ite (i = i') 1 0 := by
+  rw [LinearMap.stdBasis_apply, Function.update_apply, Pi.zero_apply]
+  congr 1; rw [eq_iff_iff, eq_comm]
+#align linear_map.std_basis_apply' LinearMap.stdBasis_apply'
+
+theorem coe_stdBasis (i : ι) : ⇑(stdBasis R φ i) = Pi.single i :=
+  rfl
+#align linear_map.coe_std_basis LinearMap.coe_stdBasis
+
+@[simp]
+theorem stdBasis_same (i : ι) (b : φ i) : stdBasis R φ i b i = b :=
+  Pi.single_eq_same i b
+#align linear_map.std_basis_same LinearMap.stdBasis_same
+
+theorem stdBasis_ne (i j : ι) (h : j ≠ i) (b : φ i) : stdBasis R φ i b j = 0 :=
+  Pi.single_eq_of_ne h b
+#align linear_map.std_basis_ne LinearMap.stdBasis_ne
+
+theorem stdBasis_eq_pi_diag (i : ι) : stdBasis R φ i = pi (diag i) := by
+  ext (x j)
+  convert(update_apply 0 x i j _).symm
+  rfl
+#align linear_map.std_basis_eq_pi_diag LinearMap.stdBasis_eq_pi_diag
+
+theorem ker_stdBasis (i : ι) : ker (stdBasis R φ i) = ⊥ :=
+  ker_eq_bot_of_injective <| Pi.single_injective _ _
+#align linear_map.ker_std_basis LinearMap.ker_stdBasis
+
+theorem proj_comp_stdBasis (i j : ι) : (proj i).comp (stdBasis R φ j) = diag j i := by
+  rw [stdBasis_eq_pi_diag, proj_pi]
+#align linear_map.proj_comp_std_basis LinearMap.proj_comp_stdBasis
+
+theorem proj_stdBasis_same (i : ι) : (proj i).comp (stdBasis R φ i) = id :=
+  LinearMap.ext <| stdBasis_same R φ i
+#align linear_map.proj_std_basis_same LinearMap.proj_stdBasis_same
+
+theorem proj_stdBasis_ne (i j : ι) (h : i ≠ j) : (proj i).comp (stdBasis R φ j) = 0 :=
+  LinearMap.ext <| stdBasis_ne R φ _ _ h
+#align linear_map.proj_std_basis_ne LinearMap.proj_stdBasis_ne
+
+theorem supᵢ_range_stdBasis_le_infᵢ_ker_proj (I J : Set ι) (h : Disjoint I J) :
+    (⨆ i ∈ I, range (stdBasis R φ i)) ≤ ⨅ i ∈ J, ker (proj i : (∀ i, φ i) →ₗ[R] φ i) := by
+  refine' supᵢ_le fun i => supᵢ_le fun hi => range_le_iff_comap.2 _
+  simp only [←ker_comp, eq_top_iff, SetLike.le_def, mem_ker, comap_infᵢ, mem_infᵢ]
+  rintro b - j hj
+  rw [proj_stdBasis_ne R φ j i, zero_apply]
+  rintro rfl
+  exact h.le_bot ⟨hi, hj⟩
+#align linear_map.supr_range_std_basis_le_infi_ker_proj LinearMap.supᵢ_range_stdBasis_le_infᵢ_ker_proj
+
+theorem infᵢ_ker_proj_le_supᵢ_range_stdBasis {I : Finset ι} {J : Set ι} (hu : Set.univ ⊆ ↑I ∪ J) :
+    (⨅ i ∈ J, ker (proj i : (∀ i, φ i) →ₗ[R] φ i)) ≤ ⨆ i ∈ I, range (stdBasis R φ i) :=
+  SetLike.le_def.2
+    (by
+      intro b hb
+      simp only [mem_infᵢ, mem_ker, proj_apply] at hb
+      rw [←
+        show (∑ i in I, stdBasis R φ i (b i)) = b by
+          ext i
+          rw [Finset.sum_apply, ← stdBasis_same R φ i (b i)]
+          refine' Finset.sum_eq_single i (fun j hjI ne => stdBasis_ne _ _ _ _ ne.symm _) _
```

Comment:
One thing that might help here is using `refine` instead and replacing only the relevant `_` with `?_`

Reply:
What would this help with?